### ISSUE TITLE
3D demo with simple building kit

### ISSUE THREAD

**Issue by mr-highball** [mr-highball/wfc#3]
spawned from [this](https://www.reddit.com/r/proceduralgeneration/comments/r2pr29/need_help_wrapping_my_head_around_wave_function/) reddit post. we'll see about handling the world generation along with attempting to get around the _invalid_ states

[image source](https://imgur.com/a/pkWeurZ)
![image](https://user-images.githubusercontent.com/35942565/143690698-d36650d3-3e44-42c5-995e-d137ead1e5d3.png)

**Comment by lynchjonathan** [comment]
Here are the meshes in one fbx
[simpleBuildingProtoSet.zip](https://github.com/mr-highball/wfc/files/7612125/simpleBuildingProtoSet.zip)

**Comment by mr-highball** [comment]
think this about sums up the constraints that I'll be following and I made [issue-3](https://github.com/mr-highball/wfc/tree/issue-3) into a branch I'll be working on 

- [x] spec out constraints
  * [commit](https://github.com/mr-highball/wfc/commit/7d96b1e955d0b0db3c8fc01606d71b68dba5edf1)
- [x] define constraints and names in wfc
  * [commit](https://github.com/mr-highball/wfc/commit/83fca7d17806610961233a82e5edb56b243d9341)
- [ ] test without 3D and see if there are any obvious faults in constraints (repeat prior steps until happy)
- [ ] convert pieces in fbx to gltf and split to files with same names
- [ ] make a demo in castle with control over parameters from a simple gui

_should_ be the full parts list with all the rotations explicitly written out. 

**basic constraints**

* ground & base parts will _share_ the same z
* mid walls will start immediately after base parts (not ground) 
* mid can also have mid pieces below or above
* top parts start above a mid parts
* top parts can have roof parts **only** above
* roof tile share the z with top parts
* roof tiles can share adjacency with other roof tiles
* roof tiles can be NORTH of roof walls
 

**naming convention**

```
gr_000 (ground)

bw_000 (base wall 0 rotation)
bw_090 (base wall 90 rotation)
bw_180 (base wall 180 rotation)
bw_270 (base wall 270 rotation)

br_000 (base corner 0 rotation)
br_090 (base corner 90 rotation)
br_180 (base corner 180 rotation)
br_270 (base corner 270 rotation)

mw_000 (mid wall 0 rotation)
mw_090 (mid wall 90 rotation)
mw_180 (mid wall 180 rotation)
mw_270 (mid wall 270 rotation)

mr_000 (mid corner 0 rotation)
mr_090 (mid corner 90 rotation)
mr_180 (mid corner 180 rotation)
mr_270 (mid corner 270 rotation)

tw_000 (top wall 0 rotation)
tw_090 (top wall 90 rotation)
tw_180 (top wall 180 rotation)
tw_270 (top wall 270 rotation)

tr_000 (top corner 0 rotation)
tr_090 (top corner 90 rotation)
tr_180 (top corner 180 rotation)
tr_270 (top corner 270 rotation)

rw_000 (roof wall 0 rotation)
rw_090 (roof wall 90 rotation)
rw_180 (roof wall 180 rotation)
rw_270 (roof wall 270 rotation)

rr_000 (roof corner 0 rotation)
rr_090 (roof corner 90 rotation)
rr_180 (roof corner 180 rotation)
rr_270 (roof corner 270 rotation)

rf_000 (roof)

em_000 (empty space block in between buildings)
```

**Comment by mr-highball** [comment]
@lynchjonathan testing out some things 1 dimension at a time and _clearly_ have some things to _work out_ (different symbols / colors mean walls/corners/grass etc.. and what I'm getting as output isn't sensible and has many invalid states). 
![image](https://user-images.githubusercontent.com/35942565/144153434-fd29c29e-fc49-43e8-b71a-b46e73fd4c5a.png)

this has been a good challenge to _poke holes_ into my implementation 😅
still working on things and will keep you posted

**Comment by lynchjonathan** [comment]
Yeah I'll need to be more explicit with the constraints between different pieces of the collapse it seems.
That on its own will make sure prototypes connect appropriately in most cases but I have gone in and implemented both a blacklisting function per face and a max and minimum height property for each prototype. Telling ground it cannot be stacked on top of ground but its okay horizontally already takes care of the stacked replica towns. 

Seems we're both getting some good feedback here too lol

Ultimately though I think the invalid state is just as valid as any other state and that is a hard rule for wfc. It was a little naive of me to think I could slap some meshes in and make it build a town. By accepting invalid as just another constraint to satisfy I can use blacklists to make sure the **perceived** invalid constraints are respected.

**Comment by mr-highball** [comment]
@lynchjonathan yes **perceived** invalid is  a great takeaway here. Conceptually it's very sensible to think that _throwing some meshes at wfc_ should "just work" but the fact is... it's not that simple haha

This is something I've been meaning to do for a while now but got sidetracked with going down a path of music generation with wfc. I've got some ideas with how to get this working properly but I want to make sure it makes sense in a _generalized_ sense, so I need to keep playing around in text to make sure it'll translate to 3D.
